javalib/src/main/scala/java/util/regex/Pattern.scala

```diff
@@ -118,18 +41,177 @@ object Pattern {
       res
     }
 
-  def adaptPatternToRe2(regex: String): String = {
-    regex
+  private object CompiledPatternStore {
+    final case class Key(regex: String, flags: Int)
+    final class Node(var key: Key,
+                     var value: RE2RegExpOps,
+                     var rc: Int,
+                     var next: Node)
+
+    private def freshNode(next: Node) =
+      new Node(null, new RE2RegExpOps(null), 0, next)
+
+    import scala.annotation.tailrec
+
+    // The tip of Nodes. The Nodes form a ring buffer of some length.
+    var last: Node = {
+      // Populate the ringbuffer
+      @tailrec def f(n: Node, num: Int): Node =
+        if (num < 0)
+          n
+        else
+          f(freshNode(n), num - 1)
+      val last = freshNode(null)
+      last.next = f(last, 128)
+      last
+    }
+
+    // Used to quickly look up a Node from a Key.
+    val map = scala.collection.mutable.HashMap.empty[Key, Node]
+
+    private def selectNode(regex: String, flags: Int): Node = synchronized {
+      // Look up a RE2RegExpOps from the map.
+      // If the map doesn't contain the key, look for an unused Node (whose refcount(rc) is 0),
+      // delete its old compiled pattern if any, and then compile a new RE2 pattern and cache it
+      // before returning it.
+      // If all of the nodes are in use, expand the ringbuffer by 1 as a last resort.
+      map.get(Key(regex, flags)).getOrElse {
+        @tailrec def findUnused(n: Node): Node =
+          if (n eq last) {
+            // No unused nodes in the ringbuffer; expand its size by 1
+            val newnode = freshNode(last.next)
+            last.next = newnode
+            newnode
+          } else if (n.rc <= 0)
+            n
+          else
+            findUnused(n.next)
+        val reused =
+          if (last.rc <= 0)
+            last
+          else
+            findUnused(last.next)
+        // delete the old pattern (if any)
+        map -= reused.key
+        if (reused.value.ptr != null) {
+          cre2.delete(reused.value.ptr)
+          reused.value = new RE2RegExpOps(null)
+        }
+        // reuse the node by replacing its members with new contents
+        reused.key = Key(regex, flags)
+        reused.value = doCompile(regex, flags)
+        map += reused.key -> reused
+        // advance `last` so that it points to the next node (which is likely the least recently used one)
+        last = reused.next
+        reused
+      }
+    }
+
+    def withRE2Regex[A](regex: String, flags: Int)(f: RE2RegExpOps => A): A = {
+      // increase the refcount of the selected node while in use to prevent it from deleted
+      val node = synchronized {
```

Comment:
`synchronized` here is probably not the best way to make this thread-safe. It's safe to assume single-threaded-only for now as we don't support threads yet.